### DISCUSSION TITLE
Extend the API to support Stateful containers

### DISCUSSION THREAD

**Discussion by aleixmorgadas** [testcontainers/testcontainers-java#7780] [Ideas]
After the [improved support for Testcontainers in the Spring Boot 3.1](https://spring.io/blog/2023/06/23/improved-testcontainers-support-in-spring-boot-3-1), I'm using the feature extensively with several DB modules such as MongoDB and PostgreSQL.

I found that I need to customize a lot the modules definition to make the containers stateful and keep the state between executions without using the `reuse` feature.

You can see an example of how I'm extending the MongoDBContainer [here](https://learnings.aleixmorgadas.dev/p/stateful-testcontainers-for-spring) in order to overcome some issues when restoring the state.

I found that Testcontainers API worked well for its initial purpose, Integration tests, but after being used for more than just test, I'm starting to experiment API gaps such as:

- Grateful container termination to keep the state consistent between runs.
  - Sometimes the data is corrupted due to containers being killed instead of stopped.
- Make it more straight forward to make a container stateful (opt-in) vs stateless (default mode).

Happy to help! :raised_hands:

**Comment by mtremsal** [comment]
@aleixmorgadas thank you for sharing the idea. What's the rationale for the following point?

> keep the state between executions **without using the `reuse` feature**.

**Reply by aleixmorgadas** [reply]
The `reuse` feature helps me to keep the state between executions while the docker is running. BUT, when the container is stopped/killed (like stopping the computer) the state is lost and I start from 0 when starting the application in dev mode again.

What I'm doing right now is adding volumes and customizing the container class, like `MongoDBContainer` to overcome some challenges when starting a container with a persisted state.

Here the problem I had with MongoDBContainer, it has a specific configuration on starting:
https://github.com/testcontainers/testcontainers-java/blob/main/modules/mongodb/src/main/java/org/testcontainers/containers/MongoDBContainer.java#L83-L87. This behavior makes the container break when the state is already there. In order to fix that, I needed to override the behavior by disabling sharding and replica set behavior. Then it worked just fine.

Same thing happens with `PostgresContainer`. Due to ungrateful shutdown, the state is corrupted and you need to change the `waitFor` strategy to enable the Postgres to recover the state.
